api/handlers.go

```diff
@@ -38,14 +39,29 @@ func (api *API) pinPOST(w http.ResponseWriter, req *http.Request, _ httprouter.P
 		return
 	}
 	// Create the skylink.
-	_, err = api.staticDB.SkylinkCreate(req.Context(), body.Skylink, conf.Conf().ServerName)
+	_, err = api.staticDB.SkylinkCreate(req.Context(), body.Skylink, api.staticConfig.ServerName)
 	if errors.Contains(err, database.ErrInvalidSkylink) {
 		api.WriteError(w, err, http.StatusBadRequest)
 		return
 	}
 	// If the skylink already exists, add this server to its list of servers.
 	if errors.Contains(err, database.ErrSkylinkExists) {
```

Comment:
Here we can technically call `SkylinkServerAdd` and leave it at that. The issue with that is that if we add other fields to the skylink record we won't be setting it there. But that that point we'll need to remove the upsert flag as well.

More specifically to your non-acid vibes, I believe the issue before was that I was doing a non-acid sequence:
1. fetch the list
2. add/remove item
3. store the list

But this time I'm not doing that and instead I'm using `update := bson.M{"$addToSet": bson.M{"servers": server}}` which should be acid.

Reply:
This is just missing distributed resource locking. By adding that you can easily make this ACID right.

Reply:
Do we need distributed resource locking here, though? We have two operations that we attempt - create and update.

The creation shouldn't need resource locking because if there is a collision we know that only one of the attempts will succeed and all others will fail, which is handled by the code.

The update should also be fine because, afaik, the `$addToSet` operation is atomic, so it doesn't need a lock.

Reply:
The problem is that you could do `SkylinkCreate` and it tells you the skylink exists.
Then you call `SkylinkFetch` but in the meantime the skylink was deleted so it will fail and it will write a `500` to the API in line 52.

Similar thing if it happens after the fetch. You might call `SkylinkMarkPinned` or `SkylinkSetverAdd` on a link that doesn't exist and return a 500 which is not ideal.

I wouldn't introduce locking though if there is a way around it. e.g. by doing a `findAndModify` with upsert or something like that.

Reply:
From offline:

We don't need the fetch if we add a flag for the `SkylinkServerAdd` method which sets the `Unpin` field to `false` if the method is being called because a user pinned a file. The alternative case for calling this method would be when performing a `sweep` action, which takes all skylinks pinned by skyd and registering them in the DB.  
It seems that removing the fetch step cleans up the code enough and makes most sense, so I'll go with that.